### PR TITLE
Xtandem-Adapter usability

### PR DESCRIPTION
**Pull request by cbielow**
- improve impl. of XTandemInfile 
    - checks XML tag ordering now via recursion
    - also empty-values parameters are loaded (to be able to preserve them when writing)

 - better docs for Xtandem-Adapter (especially on the use of the default config)

 - The param '-default_input_file' is renamed to '-default_config_file' and has a non-empty default now (the XML config in OpenMS/share). The behaviour is IDENTICAL to before, but the file is now explicitly mentioned. If the parameter is set empty, no config file will be loaded in the background (was not possible before).

 - The user can set a flag 'ignore_adapter_param', which gives precendence to the XML config file for all but four parameters: -in -out -database and -xtandem_executable.

### CODE REVIEW

**Comment by cbielow** [comment]
fixes #1379

**Comment by mwalzer** [comment]
Nice, though: would one be warned if a parameter is a doublet (with different values)? 
And: I'd find it more intuitive to have parameter precedence from the xtandem parameter file over the adapter parameters. If there already exists such a file from times using xtandem standalone for example.

File: src/openms/include/OpenMS/FORMAT/XTandemInfile.h [hunk 1]
```diff
@@ -45,7 +45,18 @@ namespace OpenMS
   /**
     @brief XTandem input file.
 
-    This class is able to create a X!Tandem configuration files to be used with Xtandem.
+    This class is able to load/write a X!Tandem configuration file.
```
Comment (hendrikweisser) [inline src/openms/include/OpenMS/FORMAT/XTandemInfile.h]: **an** X! Tandem configuration file

**Comment by hendrikweisser** [comment]
Can we please use the official spelling "X! Tandem" (with the space) everywhere? See the website: http://www.thegpm.org/tandem/

**Comment by hendrikweisser** [comment]
> would one be warned if a parameter is a doublet (with different values)? 

I agree that this would be good.

> I'd find it more intuitive to have parameter precedence from the xtandem parameter file over the adapter parameters.

This is a tricky one. We might want a parameter file to take precedence over default settings in the adapter, but if I explicitly set a parameter in the adapter, I want to be sure that my setting takes effect!
How about a flag to switch the precedence around?

**Comment by cbielow** [comment]
Am 10/30/2015 um 1:01 PM schrieb Mathias Walzer:
> Nice, though: would one be warned if a parameter is a doublet (with
> different values)?

This does not happen, since member param take precedence over whatever 
was loaded() before.

> And: I'd find it more intuitive to have parameter precedence from the
> xtandem parameter file over the adapter parameters. If there already
> exists such a file from times using xtandem standalone for example.

Nah.. bad. Since you would have to strip down the XML param file such 
that it does not overlap..or wonder why setting params in the adapter 
has no effect.

**Comment by cbielow** [comment]
Am 10/30/2015 um 1:17 PM schrieb hendrikweisser:
>     would one be warned if a parameter is a doublet (with different
>     values)?
>
> I agree that this would be good.
>
>     I'd find it more intuitive to have parameter precedence from the
>     xtandem parameter file over the adapter parameters.
>
> This is a tricky one. We might want a parameter file to take precedence
> over default settings in the adapter, but if I explicitly set a
> parameter in the adapter, I want to be sure that my setting takes effect!
> How about a flag to switch the precedence around?

Working on it.
However, I'm not going to mess around with questions like: if the 
adapter default was changed, then the value is used to override the 
default XML etc... way to complicated and not clear for the user.

There is going to be 1(!) flag: 'ignore_adapter_param'. If this is set,
the XML takes over. Otherwise ALL the adapter params are used, plus 
whatever you add to the config which is not accessible via the adapter.
End!

**Comment by hendrikweisser** [comment]
> There is going to be 1(!) flag: 'ignore_adapter_param'.

I like it.

**Comment by timosachsenberg** [comment]
The following tests FAILED:
	1743 - TOPP_XTandemAdapter_1_out (Failed)
	1746 - TOPP_XTandemAdapter_3_out (Failed)

**Comment by timosachsenberg** [comment]
+1

**Comment by lars20070** [comment]
@cbielow Thanks again for this PR. FYI, we see a significant increase in total PSMs with X!Tandem Piledriver compared to MS-GF+ in labelled and un-labelled data (about +20%).

**Comment by timosachsenberg** [comment]
Lars. Could you check if refinement is enabled? I am just asking because Julian mentioned sth. related today

**Comment by lars20070** [comment]
No, it's not. Here the complete settings I used.
https://gist.github.com/lars20070/ec38ca81cd9ff84a8a00
Please let me know, if you spot a mistake or can confirm the results in your data.

**Comment by timosachsenberg** [comment]
sounds good - to be completely certain you can check if it shows a message on the terminal (can't recall "refine", "refining" or "refinement") on execution

**Pull request merged**